sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+        neighbours_proba *= neighbours_mask
+
+        neighbours_term = np.einsum("ij,ijkl", neighbours_proba, outer)
+
+        assert neighbours_term.shape == (n_features, n_features)
+        grad -= neighbours_term
+
+        function_value = 0
+        if loss == 'l1':
+            function_value = p_i.sum()
+        elif loss == 'kl':
+            function_value = np.log(p_i).sum()
+
+        grad = 2 * np.dot(L, grad)
+
+        return [-function_value, -grad.flatten()]
+
+    return oracle
+
+
+def nca_semivectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    n_samples, n_features = X.shape
+
+    def oracle(L):
+        L = L.reshape((n_components, n_features))
+
+        Lx = np.dot(X, L.T)  # n_samples x n_components
+        assert Lx.shape == (n_samples, n_components)
+
+        grad = np.zeros((n_features, n_features))
+        function_value = 0
+
+        for i in range(n_samples):
+            A = Lx[i, :] - Lx  # n_samples x n_components
+            assert A.shape == (n_samples, n_components)
+
+            logp = -(A * A).sum(axis=1)  # n_samples
+            logp[i] = -np.inf
+            logp -= sp.misc.logsumexp(logp)
+            assert logp.shape == (n_samples, )
+
+            p = np.exp(logp)  # n_samples
+            assert p.shape == (n_samples, )
+
+            class_neighbours = y == y[i]
+            p_i = (p * class_neighbours).sum()
+
+            if loss == 'l1':
+                samples_proba = p * p_i
+            elif loss == 'kl':
+                samples_proba = p
+            mask = samples_proba > threshold  # n_samples
+
+            Xij = X[i] - X[mask, :]  # n_relevant x n_features
+            Xij_outer = (Xij[:, :, np.newaxis] * Xij[:, np.newaxis, :])
+            grad += np.einsum("i,ijk", samples_proba[mask], Xij_outer)
+
+            if loss == 'l1':
+                neighbours_proba = p * (y == y[i])
+            elif loss == 'kl':
+                neighbours_proba = p / p_i * (y == y[i])
+            mask = neighbours_proba > threshold  # n_samples
+
+            Xij = X[i] - X[mask, :]  # n_relevant x n_features
+            Xij_outer = (Xij[:, :, np.newaxis] * Xij[:, np.newaxis, :])
+            grad -= np.einsum("i,ijk", neighbours_proba[mask], Xij_outer)
+
+            if loss == 'l1':
+                function_value += p_i
+            elif loss == 'kl':
+                function_value += np.log(p_i)
+
+        grad = 2 * np.dot(L, grad)
+
+        return [-function_value, -grad.flatten()]
+
+    return oracle
+
+
+def optimize_nca(X, y, learning_rate, n_components, loss, n_init, max_iter,
+                 method, tol, random_state, cache_size, verbose):
+    n_samples, n_features = X.shape
+
+    rng = np.random.RandomState(random_state)
+    best_value = np.inf
+    best_L = None
+
+
+    threshold = 1e-3 / np.abs(X).max() ** 2
+
+    outer_products_size = (n_samples * n_features) ** 2
+    memory_permits = outer_products_size < cache_size
+    if memory_permits:
```

Comment:
Yes, definitely, it should become a keyword with fixed, direct effect
instead of cache_size which is data dependent

On Wed, Jun 3, 2015 at 8:24 AM, Vlad Niculae <notifications@github.com>
wrote:

> In sklearn/metric_learning/nca.py
> <https://github.com/scikit-learn/scikit-learn/pull/4789#discussion_r31596558>
> :
>
> > +
> > +
> > +def optimize_nca(X, y, learning_rate, n_components, loss, n_init, max_iter,
> > +                 method, tol, random_state, cache_size, verbose):
> > +    n_samples, n_features = X.shape
> > +
> > +    rng = np.random.RandomState(random_state)
> > +    best_value = np.inf
> > +    best_L = None
> > +
> > +
> > +    threshold = 1e-3 / np.abs(X).max() ** 2
> > +
> > +    outer_products_size = (n_samples * n_features) ** 2
> > +    memory_permits = outer_products_size < cache_size
> > +    if memory_permits:
>
> I absolutely agree. Cythonizing would definitely be premature. My comment
> was more on the API. If this stays relevant, I'd let the user pass
> memory_permits directly. Even more so while this is in the early
> development stage, when we want to easily benchmark the different
> approaches.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4789/files#r31596558>.
>